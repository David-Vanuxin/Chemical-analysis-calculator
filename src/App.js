import React from "react"
import {
  BrowserRouter as Router,
  Routes,
  Route,
  /*Link*/
} from "react-router-dom";

import Home from "./components/Home"
import CombustionProducts from "./components/CombustionProducts"
import QuantitativeValues from "./components/QuantitativeValues"
import Result from "./components/Result"

export default function App() {
  return (
    <Router>
      <div>
{/*        <nav>
          <ul>
            <li><Link to="/">Home</Link></li>
            <li><Link to="/counter">Counter</Link></li>
            <li><Link to="/echo">Echo</Link></li>
          </ul>
        </nav>*/}
        <Routes>
          <Route path="/" caseSensitive={false} element={<Home />} />
          <Route path="/combustion-products" caseSensitive={false} element={<CombustionProducts />} />
          <Route path="/quantitative-values" caseSensitive={false} element={<QuantitativeValues />} />
          <Route path="/result" caseSensitive={false} element={<Result />} />
        </Routes>
      </div>
    </Router>
  );
}


// Не реализованные фичи:
// Вид химических формул (нужно заключить цифру в тег <sub>) (1)
// Переход к следующему элементу <input> по нажатию enter (2)

// (1)
let times = 0
document.getElementById("root").addEventListener("DOMSubtreeModified", event => {
  console.log(`DOMtree changed ${++times} times`)
  // Лучше искать по тегу <span>
/*  const pre_labs = document.getElementsByTagName("label")

  if (pre_labs.length == 0) return

  const labels = []

  for(let i = 0; i <= (pre_labs.length / 2) + 1; ++i) labels.push(pre_labs[i])

  for (let n = 0; n < labels.length; ++n) {

    if (typeof labels[n] === "undefined") continue

    let span = labels[n].getElementsByTagName("span")[0]

    if (span.innerHTML.match(/<sub>/) != null) continue

    if (span.innerHTML.match(/\d+/) == null) continue

    let result = []
    for (let char of span.innerText) {
      if ( !isNaN( Number(char) ) ) {
        result.push(`<sub>${char}</sub>`)
      } else {
        result.push(char)
      }
    }

    result = result.join("")
    console.log(result)
    span.innerHTML = result
  }*/
})

// (2)
/*document.getElementById("root").addEventListener("DOMSubtreeModified", event => {
  let elems = document.getElementsByTagName("input")

  for (let n = 0; n < elems.length; ++n) {
    elems[n].addEventListener("keypress", (event) => {
      if (event.keyCode == 13) {
        elems[n].blur()
        if (elems[n + 1] !== "undefined") {
          elems[n + 1].focus()
          console.log("asd")
        }
      }
    })
  }
})*/