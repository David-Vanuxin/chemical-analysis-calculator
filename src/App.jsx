import React from "react"
import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom";

import * as Components from "./components"

export default function App() {
  return (
    <>
      <Router>
        <Routes>
          <Route path="/" caseSensitive={false} element={<Components.Home />} />        
          <Route path="/matter-mass" caseSensitive={false} element={<Components.MatterMass />} />
          <Route path="/combustion-products" caseSensitive={false} element={<Components.CombustionProducts />} />
          <Route path="/quantitative-values" caseSensitive={false} element={<Components.QuantitativeValues />} />
          <Route path="/result" caseSensitive={false} element={<Components.Result />} />
          {/*<Route path="/app-type" caseSensitive={false} element={<Components.AppType />} />*/}
          <Route path="/matters" caseSensitive={false} element={<Components.Matters />} />
          {/*<Route path="/percents-values" caseSensitive={false} element={<Components.PercentsValues />} />*/}
        </Routes>
      </Router>
    </>
  );
}


// Не реализованные фичи:
// использовать хуки useEffect() и useRef() в <App />, чтобы заключить цифру в тег <sub> (1)
// Переход к следующему элементу <input> по нажатию enter (2)

// (1)
/*let times = 0
document.getElementById("root").addEventListener("DOMSubtreeModified", event => {
  console.log(`DOMtree changed ${++times} times`)
  // Лучше искать по тегу <span>
  const pre_labs = document.getElementsByTagName("label")

  if (pre_labs.length == 0) return

  const labels = []

  for(let i = 0; i <= (pre_labs.length / 2) + 1; ++i) labels.push(pre_labs[i])

  for (let n = 0; n < labels.length; ++n) {

    if (typeof labels[n] === "undefined") continue

    let span = labels[n].getElementsByTagName("span")[0]

    if (span.innerHTML.match(/<sub>/) != null) continue

    if (span.innerHTML.match(/\d+/) == null) continue

    let result = []
    for (let char of span.innerText) {
      if ( !isNaN( Number(char) ) ) {
        result.push(`<sub>${char}</sub>`)
      } else {
        result.push(char)
      }
    }

    result = result.join("")
    console.log(result)
    span.innerHTML = result
  }
})*/

// (2)
/*document.getElementById("root").addEventListener("DOMSubtreeModified", event => {
  let elems = document.getElementsByTagName("input")

  for (let n = 0; n < elems.length; ++n) {
    elems[n].addEventListener("keypress", (event) => {
      if (event.keyCode == 13) {
        elems[n].blur()
        if (elems[n + 1] !== "undefined") {
          elems[n + 1].focus()
          console.log("asd")
        }
      }
    })
  }
})*/