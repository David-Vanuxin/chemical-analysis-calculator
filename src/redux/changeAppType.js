const initialState = {
  type:"values"
}

export const changeAppType = (state = initialState, action) => {
  switch(action.type) {
    case "changeAppType":
        state.type = action.value
      return state

    case "restart": return {type:"values"}

    default: return state
  }
}