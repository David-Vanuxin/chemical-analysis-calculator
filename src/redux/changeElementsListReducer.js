const initialState = {
  elements: [
    {name:"CO2", value:null},
    {name:"H2O", value:null},
  ]
}

export const changeElementsListReducer = (state = initialState, action) => {
  switch (action.type) {
    case "addElement":
      state.elements.push({
        name: action.name,
        value:null
      })
      return state

    case "removeElement":
      for (let elem of state.elements) {
        if (elem.name === action.name) {
          let index = state.elements.indexOf(elem)
          state.elements.splice(index, index + 1)
        }
      }
      return state

    case "changeMass":
      for (let elem of state.elements) {
        if (elem.name === action.name) {
          elem.value = action.value
        }
      }
      return state

    case "restart": 
      return {
        elements: [
          {name:"CO2", value:null},
          {name:"H2O", value:null},
        ]
      }

    default: return state
  }
}