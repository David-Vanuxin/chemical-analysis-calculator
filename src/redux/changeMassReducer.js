const initialState = {
  mass:null
}

export const changeMatterMassReducer = (state = initialState, action) => {
  switch(action.type) {
    case "changeMatterMass":
        state.mass = action.mass
      return state

    case "deleteMatterMass":
    case "restart": return {mass:null}

    default: return state
  }
}