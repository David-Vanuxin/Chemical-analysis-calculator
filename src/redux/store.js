import {createStore} from "redux"
import {changeElementsListReducer} from "./changeElementsListReducer"

export const store = createStore(changeElementsListReducer)