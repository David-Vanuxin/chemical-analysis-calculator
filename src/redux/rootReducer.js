import {combineReducers} from "redux"

import {changeElementsListReducer} from "./changeElementsListReducer"
import {changeMatterMassReducer} from "./changeMassReducer"
import {changeAppType} from "./changeAppType"


export const rootReducer = combineReducers({
    elementsList: changeElementsListReducer,
    matter: changeMatterMassReducer,
    type: changeAppType
  }) 