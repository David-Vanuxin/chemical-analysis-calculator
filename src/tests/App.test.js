import { render, screen } from '@testing-library/react';
import App from '../App';

import { Provider } from 'react-redux';
import {store} from "../redux/store"

import renderer from 'react-test-renderer';

describe('Начальная страница', () => {
  test('Рендерится заголовок', () => {
    render(<Provider store={store}><App /></Provider>);
    const linkElement = screen.getByText(/Калькулятор/i);
    expect(linkElement).toBeInTheDocument();
  });

  test('Рендерится кнопка', () => {
    render(<Provider store={store}><App /></Provider>);
    const linkElement = screen.getByText(/Старт/i);
    expect(linkElement).toBeInTheDocument();
  });
})


it('Рендеринг первой страницы', () => {
  const tree = renderer
    .create(<Provider store={store}><App /></Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});