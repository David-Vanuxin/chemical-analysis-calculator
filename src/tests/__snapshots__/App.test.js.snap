// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Рендеринг первой страницы 1`] = `
<div>
  <div
    className="main"
  >
    <div
      className="header-wrapper"
    >
      <h1>
        Калькулятор для химического анализа органических веществ
      </h1>
      <h2>
        По количественным характеристикам продуктов сгорания органического вещества определяет его молекулярную формулу
      </h2>
    </div>
    <div
      className="button-wrapper"
    >
      <button
        onClick={[Function]}
      >
        Старт
      </button>
    </div>
  </div>
</div>
`;
